sw/airborne/subsystems/ins/ins_ekf2.cpp

```diff
@@ -0,0 +1,509 @@
+  float rates_int[3], accel_int[3];
+  rates_int[0] = ekf2.gyro.p * ekf2.gyro_dt/1.e6f;
+  rates_int[1] = ekf2.gyro.q * ekf2.gyro_dt/1.e6f;
+  rates_int[2] = ekf2.gyro.r * ekf2.gyro_dt/1.e6f;
+  accel_int[0] = ekf2.accel.x * ekf2.accel_dt/1.e6f;
+  accel_int[1] = ekf2.accel.y * ekf2.accel_dt/1.e6f;
+  accel_int[2] = ekf2.accel.z * ekf2.accel_dt/1.e6f;
+
+  ekf.setIMUData(stamp, ekf2.gyro_dt, ekf2.accel_dt, rates_int, accel_int);
+
+  if(ekf.attitude_valid()) {
+    // Calculate the quaternion
+    struct FloatQuat ltp_to_body_quat;
+    const Quatf att_q{ekf.calculate_quaternion()};
+    ltp_to_body_quat.qi = att_q(0);
+    ltp_to_body_quat.qx = att_q(1);
+    ltp_to_body_quat.qy = att_q(2);
+    ltp_to_body_quat.qz = att_q(3);
+
+    // Publish it to the state
+    stateSetNedToBodyQuat_f(&ltp_to_body_quat);
+
+    /* Check the quaternion reset state */
+    float delta_q_reset[4];
+    uint8_t quat_reset_counter;
+    ekf.get_quat_reset(delta_q_reset, &quat_reset_counter);
+
+    // If reset update the setpoint heading
+    if(ekf2.quat_reset_counter < quat_reset_counter) {
+      float psi = matrix::Eulerf(matrix::Quatf(delta_q_reset)).psi();
+#if defined STABILIZATION_ATTITUDE_TYPE_INT
+      stab_att_sp_euler.psi += ANGLE_BFP_OF_REAL(psi);
+#else
+      stab_att_sp_euler.psi += psi;
+#endif
+      guidance_h.sp.heading += psi;
+      guidance_h.rc_sp.psi += psi;
+      nav_heading += ANGLE_BFP_OF_REAL(psi);
+      guidance_h_read_rc(autopilot_in_flight());
+      stabilization_attitude_enter();
+      ekf2.quat_reset_counter = quat_reset_counter;
+    }
+
+    /* Get in-run gyro bias */
+    struct FloatRates body_rates;
+    float gyro_bias[3];
+    ekf.get_gyro_bias(gyro_bias);
+    body_rates.p = ekf2.gyro.p - gyro_bias[0];
+    body_rates.q = ekf2.gyro.q - gyro_bias[1];
+    body_rates.r = ekf2.gyro.r - gyro_bias[2];
+
+    // Publish it to the state
+    stateSetBodyRates_f(&body_rates);
+
+    /* Get the in-run acceleration bias */
+    struct Int32Vect3 accel;
+    float accel_bias[3];
+    ekf.get_accel_bias(accel_bias);
+    accel.x = ACCEL_BFP_OF_REAL(ekf2.accel.x - accel_bias[0]);
+    accel.y = ACCEL_BFP_OF_REAL(ekf2.accel.y - accel_bias[1]);
+    accel.z = ACCEL_BFP_OF_REAL(ekf2.accel.z - accel_bias[2]);
+
+    // Publish it to the state
+    stateSetAccelBody_i(&accel);
+  }
+
+  ekf2.gyro_valid = false;
+  ekf2.accel_valid = false;
+  ekf2.got_imu_data = true;
+}
+
+/* Update INS based on Baro information */
+static void baro_cb(uint8_t __attribute__((unused)) sender_id, uint32_t stamp, float pressure)
+{
+  // Calculate the air density
+  float rho = pprz_isa_density_of_pressure(pressure, 20.0f); // TODO: add temperature compensation now set to 20 degree celcius
+  ekf.set_air_density(rho);
+  
+  // Calculate the height above mean sea level based on pressure
+  float height_amsl_m = pprz_isa_height_of_pressure_full(pressure, 101325.0); //101325.0 defined as PPRZ_ISA_SEA_LEVEL_PRESSURE in pprz_isa.h
```

Comment:
I think amsl should be calculated from QNH, not mean sea level pressure (this one gives flight level).
Actually, maybe you should get this information from the air_data module (temperature, qnh, ...)

Reply:
I will calculate the QNH for now myself, but maybe the air_data module should be updated to output ABI messages. Because I don't like adding inputs which are not coming from ABI.

Reply:
I checked the PX4 code they use the same pressure for their calculations. So I think this is fine for now and can be updated when air_data outputs abi messages. https://github.com/PX4/Firmware/blob/c415477caf00b9acdb2f3215267bccbb5b859045/src/modules/sensors/sensor_params.c#L119